parlai/agents/image_seq2seq/image_seq2seq.py

```diff
@@ -0,0 +1,349 @@
+        """
+        batch = super().batchify(obs_batch, sort)
+
+        def _process_img(img):
+            if img is not None and isinstance(img, torch.Tensor):
+                if img.dim() == 4:
+                    img = img[0, :, 0, 0]
+                if self.use_cuda:
+                    img = img.cuda()
+                if self.opt.get('fp16'):
+                    img = img.half()
+                else:
+                    img = img.float()
+
+            return img
+
+        if type(batch.image) == list and any(b is not None for b in batch):
+            images = []
+            for img in batch.image:
+                images.append(_process_img(img))
+            batch.image = images
+        return batch
+
+    def _model_input(self, batch: Batch) -> Tuple[torch.Tensor, List[object]]:
+        return (batch.text_vec, batch.image)
+
+    def load_state_dict(self, state_dict: Dict[str, torch.Tensor]):
+        """
+        Override for custom loading.
+
+        Three reasons:
+            1. When using an init model without an image encoder
+            2. When using an init model with only an encoder provided
+                In this case, we may need to add the START token to the state_dict
+            3. When using an init model without image tokens in the embeddings
+        """
+        # Case 1 -> No Image Encoder
+        if 'encoder.image_encoder.0.weight' not in state_dict:
+            for k, v in self.model.encoder.image_encoder.state_dict().items():
+                state_dict[f'encoder.image_encoder.{k}'] = v
+
+        # Case 2 -> Only an Encoder provided
+        if not (any('decoder' in state_key for state_key in state_dict)):
+            for k, v in self.model.decoder.state_dict().items():
+                state_dict[f'decoder.{k}'] = v
+            state_dict['decoder.embeddings.weight'] = state_dict['embeddings.weight']
+            if 'START' not in state_dict:
+                state_dict['START'] = self.model.START
+
+        cnt = 0
+        if self.opt['init_model'] is not None:
+            try:
+                self.model.load_state_dict(state_dict)
+                return
+            except Exception as e:
+                # Case 3 --> Check for Embedding Diffs
+                print('exception in loading state dict: {}'.format(e))
+                # make sure dims match up
+                embs = state_dict['embeddings.weight']
+                enc_embs = state_dict['encoder.embeddings.weight']
+                dec_embs = state_dict['decoder.embeddings.weight']
+                num_embs = min(self.model.embeddings.weight.shape[0], embs.shape[0])
+                num_enc_embs = min(
+                    self.model.encoder.embeddings.weight.shape[0], enc_embs.shape[0]
+                )
+                num_dec_embs = min(
+                    self.model.decoder.embeddings.weight.shape[0], dec_embs.shape[0]
+                )
+                for i in range(num_embs):
```

Comment:
i'm sorry, what is going on in all this? I see a lot of very opaque editing of embeddings and I'm unsure why we're going through these motions

Reply:
you know what, checking the cases above again, it looks like you're trying to initialize a decoder. so like, what's with the loops?

Reply:
this whole function was the one i was least comfortable with, so i tried documenting the 3 cases in the docstring. This is a case where we are initializing an **Image** TGA with an init model of a **Standard** TGA. The issue that comes up is that the new **image** TGA has two additional tokens in its dictionary - `__image__` and `__no_image__`.  As a result, the embeddings are not the right dimension. (e.g. 50,000 vs. 50,002). So, as the image tokens are presumably at the end of the dictionary, I copy over the first N-2 embedding weights from the `init_model` to the new model.

perhaps a more clear way of handling this is to check that, indeed, the embedding dimensions only differ by 2. Also, I can probably remove most of these for loops and just have one.